### PR TITLE
Fix incorrect parameter copy when passing parameters to python kernel functions

### PR DESCRIPTION
**Pull request by annagrin**
### Description

**Problem**
The structure we use for collecting arguments to pass to python is in non-packed form, but we rely on arguments to follow each other immediately (without alignment padding) when we try to find the return argument (last in that structure). This results in incorrect parameter passing to python when alignment padding is inserted (for example, in case of one bool and one int  argument).

**Solution**

Use `genReturnOffsetFunction` to create a `returnOffset` function and call it to compute the return offset for python kernels instead of performing incorrect manual calculations.

Closes: https://github.com/NVIDIA/cuda-quantum/issues/1580

### CODE REVIEW

File: lib/Optimizer/Builder/Factory.cpp [hunk 1]
```diff
@@ -73,7 +73,7 @@ cudaq::cc::StructType factory::buildInvokeStructType(FunctionType funcTy) {
     eleTys.push_back(genBufferType</*isOutput=*/false>(inTy));
   for (auto outTy : funcTy.getResults())
     eleTys.push_back(genBufferType</*isOutput=*/true>(outTy));
-  return cudaq::cc::StructType::get(ctx, eleTys /*packed=false*/);
+  return cudaq::cc::StructType::get(ctx, eleTys, true);
```
[the remark this answers is not in the document]
Comment (annagrin) [reply lib/Optimizer/Builder/Factory.cpp:76]: Here is my current understanding (please correct me if I missed something):

It is used to create a structure that contains all the arguments in `GenerateKernelExecution` : https://github.com/NVIDIA/cuda-quantum/blob/e35041833e1ea81c5ae57ea9a4fa798e7870af45/lib/Optimizer/Transforms/GenKernelExecution.cpp#L1428, and we copy the python args into it I think in `packArgs`:  https://github.com/NVIDIA/cuda-quantum/blob/e35041833e1ea81c5ae57ea9a4fa798e7870af45/python/utils/OpaqueArguments.h#L188

After the execution, we need to read the return value so we calculate where it is located in `rawArgs` (which is the structure in question as a pointer to data) in `pyAltLaunchKernelR`, assuming we know the length of all the args. This breaks with alignment padding of course: https://github.com/NVIDIA/cuda-quantum/blob/e35041833e1ea81c5ae57ea9a4fa798e7870af45/python/runtime/cudaq/platform/py_alt_launch_kernel.cpp#L269

[the remark this answers is not in the document]
Comment (annagrin) [reply lib/Optimizer/Builder/Factory.cpp]: Changed to calling `genReturnOffsetFunction` instead, thank you!

File: python/runtime/cudaq/platform/py_alt_launch_kernel.cpp [hunk 2]
```diff
@@ -183,14 +183,26 @@ jitAndCreateArgs(const std::string &name, MlirModule module,
...
+  std::size_t returnOffset = 0;
+  if (runtimeArgs.size()) {
+    auto expectedPtr = jit->lookup(name + ".returnOffset");
+    if (!expectedPtr) {
+      throw std::runtime_error(
+          "cudaq::builder failed to get returnOffset function.");
+    }
+    auto returnOffsetCreator =
+        reinterpret_cast<std::size_t (*)()>(*expectedPtr);
+    returnOffset = returnOffsetCreator();
```
Comment (schweitzpgi) [inline python/runtime/cudaq/platform/py_alt_launch_kernel.cpp:196]: Do you want to check for NoReturnOffset being returned here?

Comment (annagrin) [reply python/runtime/cudaq/platform/py_alt_launch_kernel.cpp:196]: Do you know what is returned in the case or error?

Comment (annagrin) [reply python/runtime/cudaq/platform/py_alt_launch_kernel.cpp]: Got it, updated!

File: python/runtime/cudaq/platform/py_alt_launch_kernel.cpp [hunk 3]
```diff
@@ -36,6 +36,8 @@ namespace py = pybind11;
 using namespace mlir;
 
 namespace cudaq {
+static constexpr std::int32_t NoResultOffset =
```
Comment (schweitzpgi) [inline python/runtime/cudaq/platform/py_alt_launch_kernel.cpp:39]:
```suggestion
// TODO: unify with the definition in GenKernelExec.cpp
static constexpr std::int32_t NoResultOffset =
```

File: python/runtime/cudaq/platform/py_alt_launch_kernel.cpp [hunk 4]
```diff
@@ -183,14 +185,29 @@ jitAndCreateArgs(const std::string &name, MlirModule module,
...
+          "cudaq::builder failed to get returnOffset function.");
+    }
+    auto returnOffsetCalculator =
+        reinterpret_cast<std::int64_t (*)()>(*expectedPtr);
+    returnOffset = (std::int32_t)returnOffsetCalculator();
+    if (returnOffset == NoResultOffset) {
+      returnOffset = 0;
+    }
+  }
+  return std::make_tuple(jit, rawArgs, size, returnOffset);
```
Comment (schweitzpgi) [inline python/runtime/cudaq/platform/py_alt_launch_kernel.cpp:203]:
```suggestion
  return {jit, rawArgs, size, returnOffset};
```
should work too.

**Review by schweitzpgi — APPROVED** [review]
LGTM. Thanks for hooking this up to the new function, Anna.

**Pull request merged**